src/client/index.ts

```diff
@@ -29,29 +36,103 @@ class ErrorHandler {
 }
 
 export async function launch(context: vscode.ExtensionContext): Promise<void> {
+  const isEasyProject = await isEsyProject();
+  return launchMerlinLsp(context, isEasyProject);
+}
+
+async function isEsyProject() {
+  const reasonConfig = vscode.workspace.getConfiguration("reason");
+  const forceEsy = reasonConfig.get<boolean>("forceEsy", false);
+  if (forceEsy) {
+    return true;
+  }
+
+  // TODO: we need to use workspace.workspaceFolders here and run LSP server per
+  // root. For now we'll just run LSP per workspace.
+  const root = vscode.workspace.rootPath;
+  if (root == null) {
+    return false;
+  }
+
+  const esyJson = path.join(root, "esy.json");
+  const packageJson = path.join(root, "package.json");
+  if (await exists(esyJson)) {
+    return true;
+  } else if (await exists(packageJson)) {
+    // package.json could be unrelated to esy, check if it has "esy" config
+    // then.
+    try {
+      const data = await readFile(packageJson, "utf8");
+      const json = JSON.parse(data);
+      return json.esy != null;
+    } catch (_e) {
+      return false;
+    }
+  }
+
+  return false;
+}
+
+function getMerlinLspOptions(useEsy: boolean) {
   const reasonConfig = vscode.workspace.getConfiguration("reason");
-  const module = context.asAbsolutePath(path.join("node_modules", "ocaml-language-server", "bin", "server"));
-  const options = { execArgv: ["--nolazy", "--inspect=6009"] };
-  const transport = client.TransportKind.ipc;
-  const run = { module, transport };
-  const debug = {
-    module,
-    options,
-    transport,
+  let ocamlmerlinLsp = reasonConfig.get<string | null>("path.ocamlmerlin-lsp", null);
+  if (ocamlmerlinLsp == null) {
+    ocamlmerlinLsp = isWin ? "ocamlmerlin-lsp.exe" : "ocamlmerlin-lsp";
+  }
+
+  let run;
+  if (useEsy) {
+    run = {
+      args: ["exec-command", "--include-current-env", ocamlmerlinLsp],
+      command: process.platform === "win32" ? "esy.cmd" : "esy",
+    };
+  } else {
+    run = {
+      args: [],
+      command: ocamlmerlinLsp,
+    };
+  }
+
+  const serverOptions: client.ServerOptions = {
+    debug: {
+      ...run,
+      options: {
+        env: {
+          ...process.env,
+          MERLIN_LOG: "-",
+          OCAMLRUNPARAM: "b",
+        },
+      },
+    },
+    run: {
+      ...run,
+      options: {
+        env: {
+          ...process.env,
+          MERLIN_LOG: "-",
+          OCAMLRUNPARAM: "b",
+        },
+      },
+    },
   };
-  const serverOptions = { run, debug };
+  return serverOptions;
+}
+
+export async function launchMerlinLsp(context: vscode.ExtensionContext, useEsy: boolean): Promise<void> {
```

Comment:
Let change `useEsy` arg to be `options: {useEsy: bool}` instead, this will make things more clear on call sites.